de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/NegotiationFactory.java

```diff
@@ -211,4 +207,18 @@ public AbstractIncomingProjectNegotiation newIncomingProjectNegotiation(
         throw new UnsupportedOperationException("transferType not implemented");
     }
   }
+
+  private TransferType getTransferType(ISarosSession session, JID remoteAddress) {
+    User user = session.getUser(remoteAddress);
+    if (user == null) {
+      throw new IllegalStateException("User <" + user + "> is not part of the session.");
```

Comment:
I do not know if this is the best decision (throwing exceptions). The user is allowed to leave at any time. Maybe it is better to log this issue and then return null as negotiation object and handle this properly in the Saros session manager.

Reply:
I would avoid returning null, as this is more of an unexpected behavior and the caller would need to check for it.

The correct handling in the Saros Session Manager is already difficult. For example this `return`.
https://github.com/saros-project/saros/blob/4a43f816677fdc1b92d6638ff29b3135f019d4a2/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java#L654-L657
In this case, the host user does not see an error message and the project is marked as shared, so `return` when the factory fails does not work properly.
Using an exception shows at least an error in the GUI, even if the project is marked as shared...

Since this exception handling is already broken (with IllegalArgumentException, UnsupportedOperationException) I would prefer to change the handling in `SarosSessionManager` in another PR.

Reply:
The method to set the handler is expected to be called ONCE during the whole lifetime of the application. It really makes no sense to swap it during runtime.

Reply:
@srossbach Could we please get back to the fact that IllegalArgumentException and UnsupportedOperationException are already thrown at the factory calls. I would like to change the exception handling in another PR, can we agree on that?

Reply:
Because some of you added them. I wrote this class and I do not remember that I had thrown any exceptions. Link the follow up PR and I will approve this patch.

Reply:
@srossbach #403